doc/ecmp/fine_grained_next_hop_hld.md

```diff
@@ -0,0 +1,306 @@
+    * [1.5 Warm Restart requirements ](#15-warm-restart-requirements)
+  * [2 Modules Design](#2-modules-design)
+    * [2.1 Config DB](#21-config-db)
+    * [2.2 State DB](#22-state-db)
+    * [2.3 CLI](#22-cli)
+    * [2.4 Orchestration Agent](#23-orchestration-agent)
+    * [2.5 SAI](#24-sai)
+  * [3 Flows](#3-flows)
+  * [4 Example configuration](#4-example-configuration)
+  * [5 Warm boot support](#5-warm-boot-support)
+
+###### Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 04/24/2020  |    Anish Narsian   | Initial version                   |
+
+
+# About this Manual
+This document provides the high level design for the Fine Grained ECMP feature implementation in SONiC
+Associated SAI proposal: https://github.com/opencomputeproject/SAI/blob/master/doc/ECMP/Ordered_and_Fine_Grained_ECMP.md
+# Scope
+This document describes the high level design of a staic Fine Grained ECMP feature as implemented in the application layer. 
+- In-scope: Modifying the behavior of ECMP to achieve fine grained handling of ECMP for a static set of next-hops and prefixes as defined in configuration
+- Out of scope: Dynamic ways to enable and use fine grained ECMP
+# Use case
+![](../../images/ecmp/use_case.png)
+
+Firewall or other applications running on loadbalanced VMs which maintain state of flows running through them, such that:
+- There is shared state amongst some set of firewalls so that flows can be recovered, but flow recovery is expensive so we should limit flow redistributions
+- Give that not all firewalls share state, there is a need to redistribute flows only amongst the firewalls which share state
+- An entire firewall set can go down
+
+
+# Definitions/Abbreviation
+###### Table 1: Abbreviations
+| Abbreviation             | Meaning                        |
+|--------------------------|--------------------------------|
+| ECMP                     | Equal Cost MultiPath           |
+| FG                       | Fine Grained                   |
+| NHG                      | Next Hop Group                 |
+
+
+# 1 Requirements Overview
+## 1.1 Functional requirements
+This section describes the SONiC requirements for Fine Grained ECMP next-hop groups 
+
+At a high level the following should be supported:
+
+Phase #1
+- Should be able to configure a Fine Grained ECMP group which defines a static means of redistribution of ECMP upon next-hop modifications based upon the fine grained ECMP SAI proposal
+- Standard route modifications should enable special ECMP behavior for those prefixes which desire Fine Grained ECMP. For all other prefixes the standard ECMP behavior should apply. 
+- Ability to enable consistent hashing via Fine grained ECMP for a statically defined ECMP group
+- Ability to specify a group(bank) in which ECMP redistribution should be performed out of a set of available next-hops

```

Comment:
Can you explain the notion of 'bank'. It is not explained in the SAI spec.

Reply:
Added some more details to the use case and details section. The example section covers the bank in terms of redistribution too. Let me know if you need further information on it.